source/faucet/main.d

```diff
@@ -368,6 +375,12 @@ public class Faucet : FaucetAPI
         logInfo("\tL: %s, H: %s", sutxo[0].output.value, sutxo[$-1].output.value);
 
         logInfo("\tutxo owned by Faucet: %d entries", this.state.owned_utxos.length);
+
+        auto to_freeze_pks = validators.filter!((pk) {
+            return this.state.utxos.getUTXOs(pk).byValue.all!(utxo => utxo.output.type != OutputType.Freeze) &&
+                ((pk in this.state.freeze_txs) is null || !this.randomClient().hasTransactionHash(this.state.freeze_txs[pk].hashFull));
```

Comment:
`pk !in ...`

The `hasTransactionHash` is to prevent re-sending when we just restarted, right ?

Reply:
`hasTransactionHash` is there in case the TX we sent is no longer in the pool and we still have no frozen UTXO.